modules/core/src/test/java/org/apache/ignite/cdc/ChangeDataCaptureSelfTest.java

```diff
@@ -0,0 +1,416 @@
+    @Parameterized.Parameter
+    public boolean specificConsistentId;
+
+    /** */
+    @Parameterized.Parameter(1)
+    public WALMode walMode;
+
+    /** */
+    @Parameterized.Parameters(name = "specificConsistentId={0}, walMode={1}")
+    public static Collection<?> parameters() {
+        return Arrays.asList(new Object[][] {
+            {true, WALMode.FSYNC},
+            {false, WALMode.FSYNC},
+            {true, WALMode.LOG_ONLY},
+            {false, WALMode.LOG_ONLY},
+            {true, WALMode.BACKGROUND},
+            {false, WALMode.BACKGROUND}
+        });
+    }
+
+    /** Consistent id. */
+    private UUID consistentId = UUID.randomUUID();
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        if (specificConsistentId)
+            cfg.setConsistentId(consistentId);
+
+        cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+            .setChangeDataCaptureEnabled(true)
+            .setWalMode(walMode)
+            .setWalForceArchiveTimeout(WAL_ARCHIVE_TIMEOUT)
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration().setPersistenceEnabled(true)));
+
+        cfg.setCacheConfiguration(
+            new CacheConfiguration<>(TX_CACHE_NAME).setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+        );
+
+        return cfg;
+    }
+
+    /** Simplest CDC test. */
+    @Test
+    public void testReadAllKeys() throws Exception {
+        // Read all records from iterator.
+        readAll(new UserCDCConsumer());
+
+        // Read one record per call.
+        readAll(new UserCDCConsumer() {
+            @Override public boolean onEvents(Iterator<ChangeDataCaptureEvent> evts) {
+                super.onEvents(Collections.singleton(evts.next()).iterator());
+
+                return false;
+            }
+        });
+
+        // Read one record per call and commit.
+        readAll(new UserCDCConsumer() {
+            @Override public boolean onEvents(Iterator<ChangeDataCaptureEvent> evts) {
+                super.onEvents(Collections.singleton(evts.next()).iterator());
+
+                return true;
+            }
+        });
+    }
+
+    /** */
+    private void readAll(UserCDCConsumer cnsmr) throws Exception {
+        IgniteConfiguration cfg = getConfiguration("ignite-0");
+
+        Ignite ign = startGrid(cfg);
+
+        ign.cluster().state(ACTIVE);
+
+        ChangeDataCapture cdc = new ChangeDataCapture(cfg, null, cdcConfig(cnsmr));
+
+        IgniteCache<Integer, User> cache = ign.getOrCreateCache(DEFAULT_CACHE_NAME);
+        IgniteCache<Integer, User> txCache = ign.getOrCreateCache(TX_CACHE_NAME);
+
+        addAndWaitForConsumption(
+            cnsmr,
+            cdc,
+            cache,
+            txCache,
+            ChangeDataCaptureSelfTest::addData,
+            0,
+            KEYS_CNT * 2,
```

Comment:
why KEYS_CNT * 2?

Reply:
To make sure added and removed data differs.